### PR TITLE
Added `SaveConfigCallback.save_config`

### LINKED ISSUE #14628: Upload pytorch lightning config.yaml as artifact for WandB logger
## 🚀 Feature

At this point, I don't think that the pytorch lightning `config.yaml` generated directly or with the CLI is uploaded as an artifact when using the `WandbLogger`

It would be nice to have that as an option for full reproducibility and to avoid saving out a temporary file in cases where many processes are running at the same time.

### Motivation

Saving out the `config.yaml` file for CLI usage is a great way to make things reproducible.  However, when using `WandbLogger` for storing assets/etc. you wouldn't have access to those files without manually managing them.

### Pitch

Just as you have a `log_model` option in the `WandbLogger in https://pytorch-lightning.readthedocs.io/en/stable/api/pytorch_lightning.loggers.wandb.html  what about having a `log_config` option in the constructor.  If so, it uploads a WandB artifact just like the model.  Given the importance of reproducibility, I think that defaulting this to `True` would be best.  So for example, using the CLI I could do something like
```python
    cli = LightningCLI(MyModule,run=False)
```
called with a config file like
```yaml
trainer:
  logger:
    class_path: pytorch_lightning.loggers.WandbLogger
    init_args:
      log_model: false
      log_config: true
      project: my_project_name
```
and it would log the generated config file.  Since I didn't change the `save_config_callback` maybe it still says to a file as well, but if I went `LightningCLI(MyModule,run=False,save_config_callback=None)` then it wouldn't save out a file but would log it to wandb.

As for its implementation, as a user I don't really care, but my understanding is that this could override the `save_config_callback` callback, but keep in mind that it still might be helpful to have the config file stored locally.  So my gut says that it is best if this is ortthogonal to that callback, so users can still turn on/off the file logging or swap out their on `save_config_callback`.  But none of my business what happens behind the scenes.  

One other thing.  In the assets, wandb has its own `config.yaml` stored in the artifacts which is confusing, so the filename would need to be something different, or maybe even customizable.  e.g. `log_config = None`` vs. `log_config = "my_filename_in_artifacts.yaml"` etc.

### Alternatives

I tried to setup the code to upload the content directly for the run but it was just too messy.  Focusing on the LIghtningCLI note that if I have
```python
    cli = LightningCLI(MyModule,run=False,save_config_callback=None)
```
with a config file like
```yaml
trainer:
  logger:
    class_path: pytorch_lightning.loggers.WandbLogger
    init_args:
      dir: .wandb
      log_model: false
      project: my_project_name
      tags:
        - my_tag
```
Then at least it won't save a file in the `.wandb` directory - which leads to clashes with multiplie concurrent processes.

I imagine this could be done by implementing a `save_config_callback` manually, but I think it might be a lot of boilerplate.

cc @borda @awaelchli @edward-io @ananthsub @rohitgr7 @Blaizzy @carmocca @mauvilsa

### PR DESCRIPTION
**Pull request by mauvilsa**
## What does this PR do?

Makes it easy to subclass `SaveConfigCallback` to additionally save the config in other places, such as a logger. The main purpose of this feature is to provide a way to implement a custom save config without having to worry about ranks or race conditions.

Motivated by https://github.com/Lightning-AI/lightning/issues/15340#issuecomment-1517274116

Might fix #14628, unless we consider that to be an out-of-the-box saving as artifact in logger.

<details>
  <summary><b>Before submitting</b></summary>

- [x] Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

### CODE REVIEW

File: src/lightning/pytorch/cli.py [hunk 1]
```diff
@@ -228,6 +228,17 @@ def setup(self, trainer: Trainer, pl_module: LightningModule, stage: str) -> Non
...
+        if trainer.is_global_zero:
+            # save only on rank zero to avoid race conditions.
+            self.extra_save_config(trainer, pl_module, stage)
+            self.already_saved = True
+
+        # broadcast so that all ranks are in sync on future calls to .setup()
+        self.already_saved = trainer.strategy.broadcast(self.already_saved)
+
+    def save_config(self, trainer: Trainer, pl_module: LightningModule, stage: str) -> None:
+        """Saves the config to the log_dir only on rank zero."""
```
Comment (carmocca) [inline src/lightning/pytorch/cli.py:241]: Why this refactor instead of just adding `self.extra_save_config()` after `self.parser.save()` in the original implementation?

File: docs/source-pytorch/cli/lightning_cli_advanced.rst [hunk 2]
```diff
@@ -60,6 +60,26 @@ with ``save_config_callback=None``.
...
+                if isinstance(trainer.logger, Logger):
+                    config = self.parser.dump(self.config)  # Required for proper reproducibility
+                    trainer.logger.log_hyperparams({"config": config})
+
+
+        cli = LightningCLI(..., save_config_callback=LoggerSaveConfigCallback)
+
+-- note::
+
+    The ``extra_save_config`` method is only run on rank zero. If you need to broadcast to other ranks, then implement
```
Comment (carmocca) [inline docs/source-pytorch/cli/lightning_cli_advanced.rst:79]: Maybe `rank_zero_save_config` would be a clearer name

File: src/lightning/pytorch/cli.py (hunk 1 above)
Comment (mauvilsa) [reply src/lightning/pytorch/cli.py:241]: Could be done differently. This is just my first idea given broadcasts in the code which I don't fully know how to deal with yet. Two main objectives are:
- Allow users to implement a custom save without having to worry about broadcast.
- Have the choice to keep or not the standard save to `log_dir`.

Another idea I had was with a single method, say `save_config`. It would be called from `setup` inside the `if trainer.is_global_zero:` block. If someone subclasses `SaveConfigCallback` and implements `save_config`, they would need to call `super().save_config` to preserve the standard save to `log_dir`. The problem is that the standard save broadcasts `log_dir` and `file_exists`, which means that it can't be called from inside the `if trainer.is_global_zero:`. We would need a standard save that does not require any broadcast.

Any ideas?

Comment (carmocca) [reply src/lightning/pytorch/cli.py:241]: For customizing the `log_dir`, I would suggest a simple `log_dir: Optional[str]` argument in the callback init

Then in setup it does:

```python
if self.log_dir is None:
    log_dir = trainer.log_dir  # this broadcast the directory
else:
    log_dir = broadcast(log_dir)
```

I think the `save_config` idea is good, perhaps with an `on_after_save_config` hook that replaces your `save_config_extra` idea


```python
def save_config(self, config_path):
    self.parser.save(
        self.config, config_path, skip_none=False, overwrite=self.overwrite, multifile=self.multifile
    )


...

        # save the file on rank 0
        if trainer.is_global_zero:
            fs.makedirs(log_dir, exist_ok=True)
            self.save_config(config_path)
            self.on_after_save_config(config_path)
            self.already_saved = True
```

Comment (mauvilsa) [reply src/lightning/pytorch/cli.py:241]: The signature for `save_config` needs to be

```python
def save_config(self, trainer: Trainer, pl_module: LightningModule, stage: str) -> None:
```

Otherwise it wouldn't be possible extend to save in a logger or for other use cases people might come up with.

To disable the standard save, an init parameter could be, but like  `save_to_log_dir: bool = True`. Though it isn't an ideal solution. It is less simple to disable. In the subclass you would need to implement `__init__` to hard code `super().__init__(*args, save_to_log_dir=False, **kwargs)` or provide this when instantiating the LightningCLI class.

Comment (mauvilsa) [reply src/lightning/pytorch/cli.py]: I pushed a new implementation for this.

File: docs/source-pytorch/cli/lightning_cli_advanced.rst [hunk 3]
```diff
@@ -60,6 +60,30 @@ with ``save_config_callback=None``.
...
+        class LoggerSaveConfigCallback(SaveConfigCallback):
+            def save_config(self, trainer: Trainer, pl_module: LightningModule, stage: str) -> None:
+                if isinstance(trainer.logger, Logger):
+                    config = self.parser.dump(self.config, skip_none=False)  # Required for proper reproducibility
+                    trainer.logger.log_hyperparams({"config": config})
+
+
+        cli = LightningCLI(..., save_config_callback=LoggerSaveConfigCallback)
+
+-- tip::
```
Comment (carmocca) [inline docs/source-pytorch/cli/lightning_cli_advanced.rst:77]:
```suggestion
.. tip::
```

File: src/lightning/pytorch/cli.py [hunk 4]
```diff
@@ -216,50 +217,71 @@ def __init__(
...
-        # save the file on rank 0
         if trainer.is_global_zero:
-            # save only on rank zero to avoid race conditions.
-            # the `log_dir` needs to be created as we rely on the logger to do it usually
-            # but it hasn't logged anything at this point
-            fs.makedirs(log_dir, exist_ok=True)
-            self.parser.save(
-                self.config, config_path, skip_none=False, overwrite=self.overwrite, multifile=self.multifile
-            )
+            self.save_config(trainer, pl_module, stage)
```
Comment (carmocca) [inline src/lightning/pytorch/cli.py:269]: Since `save_config` is called only on global zero, any collective call inside will make it hang.
A simple example being that a users calls `trainer.log_dir`.

This would be very limiting when `save_to_log_dir is False`. Why don't we let it run on all ranks?

File: src/lightning/pytorch/cli.py [hunk 5]
```diff
@@ -216,50 +217,71 @@ def __init__(
...
             self.already_saved = True
 
         # broadcast so that all ranks are in sync on future calls to .setup()
         self.already_saved = trainer.strategy.broadcast(self.already_saved)
 
+    def save_config(self, trainer: Trainer, pl_module: LightningModule, stage: str) -> None:
+        """Implement to save the config in some other place different from the standard log_dir.
```
Comment (carmocca) [inline src/lightning/pytorch/cli.py:276]: The wording here suggests that if you implement this hook, then config doesn't get saved to `log_dir`. That's not true unless you also pass `save_to_log_dir=False`.

**Review by carmocca — COMMENTED** [review]
This looks better than before but still doesn't seem ideal to me. Left comments!

File: src/lightning/pytorch/cli.py (hunk 4 above)
Comment (carmocca) [reply src/lightning/pytorch/cli.py:269]: > Is there any way (e.g. context manager) to make it such that if someone tries to make a collective call, it simply fails instead of hanging?

Not at the moment.

Resolving

File: src/lightning/pytorch/cli.py (hunk 5 above)
Comment (mauvilsa) [reply src/lightning/pytorch/cli.py]: Changed it. Also added a note about the hanging.

**Review by Borda — APPROVED** [review]

File: docs/source-pytorch/cli/lightning_cli_advanced.rst [hunk 6]
```diff
@@ -60,6 +60,36 @@ with ``save_config_callback=None``.
...
+It is also possible to extend the :class:`~lightning.pytorch.cli.SaveConfigCallback` class, for instance to additionally
+save the config in a logger. An example of this is:
+
+    .. code:: python
+
+        class LoggerSaveConfigCallback(SaveConfigCallback):
+            def save_config(self, trainer: Trainer, pl_module: LightningModule, stage: str) -> None:
+                if isinstance(trainer.logger, Logger):
+                    config = self.parser.dump(self.config, skip_none=False)  # Required for proper reproducibility
+                    trainer.logger.log_hyperparams({"config": config})
```
Comment (mauvilsa) [inline docs/source-pytorch/cli/lightning_cli_advanced.rst:72]: @tommycwh I answer your [comment](https://github.com/Lightning-AI/lightning/issues/15340#issuecomment-1532397297) since it is about this pull request.

> However, I would want some more instructions on how to implement my save_config() function. For example, in my example above, there are some hyperparameters (like those numerical values) that can be saved by the "default" config saving mechanism and I would need to implement my own mechanism for the others that cannot be saved automaically (like the init_args).

I am not sure if I truly understand what you mean or what exactly is it that you want to achieve. This pull request adds `save_config` such that it can be used however people want to use it. In principle the config is intended for reproducibility and I would save it in a logger as an artifact or a single yaml string which includes any `init_args`. This is just three lines of code like shown in the example. Doing something different is perfectly fine, but it depends on what you want to do.

**Review by carmocca — APPROVED** [review]
